### ISSUE TITLE
Practical application: Could this be used to replace an MPC controller from Acado Toolkit?

### ISSUE THREAD

**Issue by zorrobyte** [locuslab/mpc.pytorch#17]
Heyo!

I'm a user and a layman contributor to OpenPilot, an open source driving agent. With some python and a consumer phone, I've added enhanced lane keep assist to my Toyota Prius for under $500 USD.
https://github.com/commaai/openpilot

I spend my time tuning lateral control and have found their plant model (vehicle_model.py) to be insufficient to properly estimate the path of the vehicle with their MPC solution they've written in Acado Toolkit.

You can find their MPC here:
https://github.com/commaai/openpilot/tree/devel/selfdrive/controls/lib/lateral_mpc

And the vehicle model here:
https://github.com/commaai/openpilot/blob/devel/selfdrive/controls/lib/vehicle_model.py

PID or INDI control (user selectable) is used to send torque to the vehicle's EPS motor. The input to PID is desired angle, which sends torque commands to drive the wheels to the desired angle. (PID) https://github.com/commaai/openpilot/blob/devel/selfdrive/controls/lib/pid.py

The existing MPC controller puts these arrays out every cycle:
![image0 png](https://user-images.githubusercontent.com/7257172/59807754-437e2b00-92c7-11e9-9d26-138ef348e24f.jpeg)

Lastly, the inputs/outputs of MPC are as follows:
self.libmpc = libmpc_py.libmpc
    self.libmpc.init(MPC_COST_LAT.PATH, MPC_COST_LAT.LANE, MPC_COST_LAT.HEADING, steer_rate_cost)

    self.mpc_solution = libmpc_py.ffi.new("log_t *")
    self.cur_state = libmpc_py.ffi.new("state_t *")
    self.cur_state[0].x = 0.0
    self.cur_state[0].y = 0.0
    self.cur_state[0].psi = 0.0
    self.cur_state[0].delta = 0.0

    self.angle_steers_des = 0.0
    self.angle_steers_des_mpc = 0.0
    self.angle_steers_des_prev = 0.0
    self.angle_steers_des_time = 0.0

In which they reference these values:
states[0].x = v_ego (speed)
  states[0].psi = v_ego (speed) * curvature_factor (how curvy is the road center) * math.radians(steer_angle) / steer_ratio
  states[0].delta = math.radians(steer_angle) / steer_ratio

Long story short, and what I'm assuming is clearly out of your scope as this project's maintainer, can mpc.pytoch be used as a replacement for MPC exported from acado toolkit? If I'm understanding the readme correctly,
"and non-linear system transition dynamics that can be defined by hand if you understand your environment or a neural network if you don’t."

Does this mean that mpc.pytorch can, um, use a neural network to learn the plant model?

Apologies if I'm off track in any of this, I'm not formally educated and knew nothing about MPC, self driving or any of this a few months ago :) I think it could be a good real world use case for your library and I'm beginning to learn just enough to be dangerous!

**Comment by bamos** [comment]
You can use this for solving a nonconvex control problem with neural network dynamics, but it's susceptible to hitting bad local optima and may not be the best approach. A few papers lately [like this one](https://arxiv.org/pdf/1805.12114.pdf) have reported good results recently using the cross-entropy method with neural network dynamics. That paper has a tensorflow implementation of the CEM for control here: https://github.com/kchua/handful-of-trials/blob/master/dmbrl/controllers/MPC.py
